services/src/ogc/wfs/request.rs

```diff
@@ -0,0 +1,193 @@
+pub struct GetPropertyValue {
+    // TODO
+}
+
+#[derive(PartialEq, Debug, Deserialize, Serialize)]
+pub struct GetFeatureWithLock {
+    // TODO
+}
+
+#[derive(PartialEq, Debug, Deserialize, Serialize)]
+pub struct CreateStoredQuery {
+    // TODO
+}
+
+#[derive(PartialEq, Debug, Deserialize, Serialize)]
+pub struct DropStoredQuery {
+    // TODO
+}
+
+#[derive(PartialEq, Debug, Deserialize, Serialize)]
+pub struct ListStoredQueries {
+    // TODO
+}
+
+#[derive(PartialEq, Debug, Deserialize, Serialize)]
+pub struct DescribeStoredQueries {
+    // TODO
+}
+
+#[allow(clippy::option_if_let_else)]
+pub fn parse_type_names<'de, D>(deserializer: D) -> Result<TypeNames, D::Error>
+where
+    D: serde::Deserializer<'de>,
+{
+    let s = <&str as serde::Deserialize>::deserialize(deserializer)?;
+
+    if let Some(pos) = s.find(':') {
+        let namespace = Some(s[..pos].to_string());
+        let feature_type = s[pos + 1..].to_string();
+
+        Ok(TypeNames {
+            namespace,
+            feature_type,
+        })
+    } else {
+        Ok(TypeNames {
+            namespace: None,
+            feature_type: s.into(),
+        })
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use geoengine_datatypes::primitives::Coordinate2D;
+    use geoengine_datatypes::projection::ProjectionAuthority;
+
+    #[test]
+    fn deserialize_get_feature() {
+        let query = "request=GetFeature&service=WFS&version=2.0.0&typeNames=ns:test&bbox=1,2,3,4";
+        let parsed: WFSRequest = serde_urlencoded::from_str(query).unwrap();
+
+        let request = WFSRequest::GetFeature(GetFeature {
+            version: "2.0.0".into(),
+            time: None,
+            srs_name: None,
+            namespaces: None,
+            count: None,
+            sort_by: None,
+            result_type: None,
+            filter: None,
+            bbox: BoundingBox2D::new(Coordinate2D::new(1., 2.), Coordinate2D::new(3., 4.)).unwrap(),
+            type_names: TypeNames {
+                namespace: Some("ns".into()),
+                feature_type: "test".into(),
+            },
+            property_name: None,
+        });
+
+        assert_eq!(parsed, request);
+    }
+
+    #[test]
+    fn deserialize_get_feature_full() {
+        let query = "request=GetFeature&service=WFS&version=2.0.0&typeNames=ns:test&bbox=1,2,3,4&srsName=EPSG:4326&format=image/png&time=2000-01-01T00:00:00.0Z/2000-01-02T00:00:00.0Z&namespaces=xmlns%28dog%3Dhttp%3A%2F%2Fwww.example.com%2Fnamespaces%2Fdog%29&count=10&sortBy=Name%5B%2BA%5D&resultType=results&filter=%3CFilter%3E%0A%20%20%3CAnd%3E%0A%20%20%20%20%3CPropertyIsEqualTo%3E%3CValueReference%3Edog%3Aage%3C%2FValueReference%3E%3CLiteral%3E2%3C%2FLiteral%3E%3C%2FPropertyIsEqualTo%3E%0A%20%20%20%20%3CPropertyIsEqualTo%3E%3CValueReference%3Edog%3Aweight%3C%2FValueReference%3E%3CLiteral%3E5%3C%2FLiteral%3E%3C%2FPropertyIsEqualTo%3E%0A%20%20%3C%2FAnd%3E%0A%3C%2FFilter%3E&propertyName=P1,P2";
```

Comment:
I guess we need a way to make this better to understand?
Maybe write query and then urlencode later?
Or maybe puzzle together from parts?

Reply:
f600add221e9e37f2faa4f90cd81bd138a278df5